actionpack/lib/action_dispatch/middleware/stack.rb

```diff
@@ -169,19 +173,31 @@ def build(app = nil, &block)
     end
 
     private
+      def nested_middleware?(klass)
+        klass.respond_to?(:nested_middleware?) && klass.nested_middleware?
+      end
+
       def assert_index(index, where)
         i = index.is_a?(Integer) ? index : index_of(index)
         raise "No such middleware to insert #{where}: #{index.inspect}" unless i
         i
       end
 
       def build_middleware(klass, args, block)
-        Middleware.new(klass, args, block)
+        if nested_middleware?(klass)
+          klass
```

Comment:
Would it make sense to build something like a `MiddlewareContainer` with the proxy here, and have that class be responsible for the nested middleware flattening?